grails-app/services/org/pih/warehouse/inventory/RequisitionStatusTransitionEventService.groovy

```diff
@@ -9,40 +9,30 @@
 **/
 package org.pih.warehouse.inventory
 
-import org.codehaus.groovy.grails.commons.GrailsApplication
-import org.pih.warehouse.core.MailService
 import org.pih.warehouse.core.Person
+import org.pih.warehouse.report.NotificationService
 import org.pih.warehouse.requisition.Requisition
 import org.pih.warehouse.requisition.RequisitionService
 import org.pih.warehouse.requisition.RequisitionStatusTransitionEvent
+import org.pih.warehouse.requisition.RequisitionStatus
 import org.springframework.context.ApplicationEvent
 import org.springframework.context.ApplicationListener
 
 class RequisitionStatusTransitionEventService implements ApplicationListener<RequisitionStatusTransitionEvent> {
 
     static transactional = true
 
-    MailService mailService
-    GrailsApplication grailsApplication
     RequisitionService requisitionService
+    NotificationService notificationService
 
 
-    void publishDefaultEmailNotifications(Requisition requisition, List<Person> receivers) {
-        def g = grailsApplication.mainContext.getBean('org.codehaus.groovy.grails.plugins.web.taglib.ApplicationTagLib')
-        String subject = "${requisition.requestNumber} ${requisition.name}"
-        String redirectToRequestsList = "${g.createLink(uri: "/stockMovement/list?direction=OUTBOUND&sourceType=ELECTRONIC", absolute: true)}"
-        GString body = "${g.render(template: "/email/approvalsAlert", model: [requisition: requisition, redirectUrl: redirectToRequestsList])}"
-
-        receivers.each {receiver ->
-            if (receiver.email) {
-                mailService.sendHtmlMail(subject, body, receiver.email)
-            }
-        }
-    }
-
     void onApplicationEvent(RequisitionStatusTransitionEvent event) {
+        if (event.newStatus == RequisitionStatus.PENDING_APPROVAL) {
+            Requisition req = event.requisition
+            List<Person> approvers = Person.findAllByIdInList(event.requisition.approvers.collect {it.id})
```

Comment:
When looping through users when sending the notification I was getting a LazyInitialization exception, which is why I needed to fetch the approvers list separately.

This looks a bit weird to me but I didn't find any other approach to fetch approvers eagerly, if you have a suggestion then please let me know.

Reply:
You can follow what I've done for the same problem with requisition events in my draft PR #4288  - you can add `fetchMode: 'join'` in Requisition mapping in this place:
https://github.com/openboxes/openboxes/blob/98806dcf18313fa6b2d27fb78641361a69bd0d0a/grails-app/domain/org/pih/warehouse/requisition/Requisition.groovy#L154

It will `left join` approvers when fetching a Requisition and won't cause an additional query to fetch the approvers, as you've currently done.

Normally, I would not recommend doing this, since it kinda adds fetch mode eager for the Requisition, but since we'll migrate to Grails 3 in a moment, I've just left a comment above mapping, that it should be removed and replaced by a proper data service get method with joining those.

Data services are not available in Grails 1, so I think it is not worth to "fight" with that Lazy in a "pattern" way, since we'll refactor that in a moment.
Remember to add a comment to refactor this in Grails 3 - the same, as I've added for events.